sway/main.c

```diff
@@ -194,21 +194,23 @@ static void log_kernel(void) {
 }
 
 
-static void drop_permissions(void) {
+static bool drop_permissions(void) {
 	if (getuid() != geteuid() || getgid() != getegid()) {
 		if (setgid(getgid()) != 0) {
-			wlr_log(WLR_ERROR, "Unable to drop root");
-			exit(EXIT_FAILURE);
+			wlr_log(WLR_ERROR, "Unable to drop root, refusing to start");
+			return false;
 		}
 		if (setuid(getuid()) != 0) {
-			wlr_log(WLR_ERROR, "Unable to drop root");
-			exit(EXIT_FAILURE);
+			wlr_log(WLR_ERROR, "Unable to drop root, refusing to start");
+			return false;
 		}
 	}
 	if (setuid(0) != -1) {
-		wlr_log(WLR_ERROR, "Root privileges can be restored.");
-		exit(EXIT_FAILURE);
+		wlr_log(WLR_ERROR, "Unable to drop root (root privileges can be "
+			"restored, and shouldn't), refusing to start");
```

Comment:
"Unable to drop root - we shouldn't be able to restore it after setuid - refusing to start"

Reply:
Done